### PR TITLE
PVS-Studio: fixed weaknesses CWE-476 (NULL Pointer Dereference)

### PR DESCRIPTION
**Pull request by Pasick**
We have found and fixed vulnerabilities CWE-476 (NULL Pointer Dereference) using PVS-Studio tool.
Analyzer warnings: [V3019](https://www.viva64.com/en/w/V3019/), [V3042](https://www.viva64.com/en/w/V3042/) and [V3125](https://www.viva64.com/en/w/V3125/).
PVS-Studio is a static code analyzer for C, C++ and C#.

### CODE REVIEW

File: src/VisualStudio/Core/Def/Implementation/Options/RoamingVisualStudioProfileOptionPersister.cs [hunk 1]
```diff
@@ -230,7 +230,7 @@ public bool TryPersist(OptionKey optionKey, object value)
                 // We store these as strings, so serialize
                 var valueToSerialize = value as CodeStyleOption<bool>;
 
-                if (value != null)
+                if (valueToSerialize != null)
```
Comment (jmarolf) [inline src/VisualStudio/Core/Def/Implementation/Options/RoamingVisualStudioProfileOptionPersister.cs]: instead of this:
```C#
var valueToSerialize = value as CodeStyleOption<bool>;	  
if (valueToSerialize != null)
{
    value = valueToSerialize.ToXElement().ToString();
}
```

could we do this?:
```C#
if (value is CodeStyleOption<bool> valueToSerialize )
{
    value = valueToSerialize.ToXElement().ToString();
}
```

File: src/VisualStudio/Core/Def/Implementation/Options/RoamingVisualStudioProfileOptionPersister.cs [hunk 2]
```diff
@@ -240,7 +240,7 @@ public bool TryPersist(OptionKey optionKey, object value)
                 // We store these as strings, so serialize
                 var valueToSerialize = value as NamingStylePreferences;
 
-                if (value != null)
+                if (valueToSerialize != null)
```
Comment (jmarolf) [inline src/VisualStudio/Core/Def/Implementation/Options/RoamingVisualStudioProfileOptionPersister.cs]: instead of this:
```C#
var valueToSerialize = value as NamingStylePreferences;	  
if (valueToSerialize != null)
{
    value = valueToSerialize.ToXElement().ToString();
}
```

could we do this?:
```C#
if (value is NamingStylePreferences valueToSerialize )
{
    value = valueToSerialize.ToXElement().ToString();
}
```

File: src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs [hunk 3]
```diff
@@ -317,8 +317,10 @@ public static bool IsInOutContext(this ExpressionSyntax expression)
         }
 
         public static bool IsInRefContext(this ExpressionSyntax expression)
-            => expression.IsParentKind(SyntaxKind.RefExpression) ||
-               (expression?.Parent as ArgumentSyntax)?.RefOrOutKeyword.Kind() == SyntaxKind.RefKeyword;
+        {
+            return expression != null && (expression.IsParentKind(SyntaxKind.RefExpression) ||
```
Comment (jmarolf) [inline src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs]:
```C#
=> expression?.IsParentKind(SyntaxKind.RefExpression) == true || 
 (expression?.Parent as ArgumentSyntax)?.RefOrOutKeyword.Kind() == SyntaxKind.RefKeyword;
```

File: src/Workspaces/Core/Desktop/Workspace/Storage/PersistentStorageService.cs [hunk 4]
```diff
@@ -112,7 +112,7 @@ private IPersistentStorage GetStorage(Solution solution, string workingFolderPat
                     // everything seems right, use what we have
                     if (storage?.WorkingFolderPath == workingFolderPath)
                     {
-                        storage.AddRefUnsafe();
+                        storage?.AddRefUnsafe();
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Desktop/Workspace/Storage/PersistentStorageService.cs]: I believe the implication is this is unneeded: the if check could only practically succeed if workingFolderPath was non-null.

File: src/Workspaces/Remote/Core/Services/SolutionService.cs [hunk 5]
```diff
@@ -146,14 +146,14 @@ private static Solution GetAvailableSolution(Checksum solutionChecksum)
             if (currentSolution?.Item1 == solutionChecksum)
             {
                 // asked about primary solution
-                return currentSolution.Item2;
+                return currentSolution?.Item2;
```
Comment (jasonmalinowski) [inline src/Workspaces/Remote/Core/Services/SolutionService.cs]: Unneeded: if the == matched then we know it's not null in practice.

File: src/Workspaces/Remote/Core/Services/SolutionService.cs [hunk 6]
```diff
@@ -146,14 +146,14 @@ private static Solution GetAvailableSolution(Checksum solutionChecksum)
...
             }
 
             var lastSolution = s_lastSolution;
             if (lastSolution?.Item1 == solutionChecksum)
             {
                 // asked about last solution
-                return lastSolution.Item2;
+                return lastSolution?.Item2;
```
Comment (jasonmalinowski) [inline src/Workspaces/Remote/Core/Services/SolutionService.cs]: Unneeded: if the == matched then we know it's not null in practice.

File: src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs [hunk 7]
```diff
@@ -2166,7 +2168,7 @@ private static bool IsThisOrTypeOrNamespace(MemberAccessExpressionSyntax memberA
 
             var initializerType = semanticModel.GetTypeInfo(equalsValueClause.Value, cancellationToken).Type;
 
-            if (!type.Equals(initializerType))
+            if (type != null && !type.Equals(initializerType))
```
Comment (jasonmalinowski) [inline src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs]: I just prefer `object.Equals(type, initializerType)`.

File: src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs (hunk 3 above)
Comment (jasonmalinowski) [inline src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs]: Let's just break this into multiple lines.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs [hunk 8]
```diff
@@ -954,7 +954,7 @@ internal void UpdateDocumentContextIfContainsDocument(IVsHierarchy sharedHierarc
                 return;
             }
 
-            if (hostProject.Id != documentId.ProjectId)
+            if (hostProject?.Id != documentId.ProjectId)
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs]: @dpoeschl What's the right thing here?

File: src/Workspaces/Core/Desktop/Workspace/Storage/PersistentStorageService.cs (hunk 4 above)
Comment (Pasick) [inline src/Workspaces/Core/Desktop/Workspace/Storage/PersistentStorageService.cs]: @jasonmalinowski, I agree, the `PersistentStorageService.GetStorage` method is called from one place only, and there's a check that `workingFolderPath` is not null, so probably this check is excessive. However, if someone will pass a null as a `workingFolderPath` argument, it could cause a NullReferenceException,

File: src/Workspaces/Remote/Core/Services/SolutionService.cs (hunk 5 above)
Comment (Pasick) [inline src/Workspaces/Remote/Core/Services/SolutionService.cs]: @jasonmalinowski, if solutionChecksum is null, we may get the NullReferenceException, so I believe that this extra check is useful.

File: src/Workspaces/Remote/Core/Services/SolutionService.cs (hunk 6 above)
Comment (Pasick) [inline src/Workspaces/Remote/Core/Services/SolutionService.cs]: @jasonmalinowski, if solutionChecksum is null, we may get the NullReferenceException, so I believe that this extra check is useful.

**Comment by CyrusNajmabadi** [comment]
@Pasick Can you give more information?  Were you able to actually get null ref exceptions in these codepaths, or are you just speculating that it is possible?

File: src/Workspaces/Remote/Core/Services/SolutionService.cs (hunk 5 above)
Comment (jasonmalinowski) [inline src/Workspaces/Remote/Core/Services/SolutionService.cs]: If solutionChecksum is null, I want it crashing.

File: src/Workspaces/Core/Desktop/Workspace/Storage/PersistentStorageService.cs (hunk 4 above)
Comment (jasonmalinowski) [inline src/Workspaces/Core/Desktop/Workspace/Storage/PersistentStorageService.cs]: Cool, it should crash in that case. :-)

**Comment by Pasick** [comment]
@CyrusNajmabadi No, I didn't get null reference exceptions in these pieces of code. I just inspected the code with a static code analyzer PVS-Studio, analyzed its warnings and decided to fix potential issues in code which may cause null reference exceptions.

File: src/Workspaces/Remote/Core/Services/SolutionService.cs (hunk 5 above)
Comment (Pasick) [inline src/Workspaces/Remote/Core/Services/SolutionService.cs]: OK, got it.

File: src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs [hunk 9]
```diff
@@ -2166,7 +2166,7 @@ private static bool IsThisOrTypeOrNamespace(MemberAccessExpressionSyntax memberA
 
             var initializerType = semanticModel.GetTypeInfo(equalsValueClause.Value, cancellationToken).Type;
 
-            if (!type.Equals(initializerType))
+            if (!Equals(type, initializerType))
```
Comment (mavasani) [inline src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs]: We should bail out and return true if either `type` or `initializerType` is non-null. @Pasick I presume your analyzer flags this because there was a null check done on type earlier?

File: src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs [hunk 10]
```diff
@@ -317,7 +317,7 @@ public static bool IsInOutContext(this ExpressionSyntax expression)
         }
 
         public static bool IsInRefContext(this ExpressionSyntax expression)
-            => expression.IsParentKind(SyntaxKind.RefExpression) ||
+            => expression?.IsParentKind(SyntaxKind.RefExpression) == true ||
```
Comment (mavasani) [inline src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs]: All of our `IsParentKind` extensions have a null check upfront and we avoid null checks at all call sites. Please revert this change.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs (hunk 8 above)
Comment (mavasani) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs]: Ping @dpoeschl to suggest the correct logic here.

**Comment by jinujoseph** [comment]
@Pasick  closing this PR for now. Feel free to reopen when you get a chance to work on this again.
